modules/src/erc721token.rs

```diff
@@ -0,0 +1,532 @@
+use crate::erc721::erc721::events::Transfer;
+use odra::contract_env::{caller, revert};
+use odra::types::event::OdraEvent;
+use odra::types::Address;
+use odra::types::U256;
+
+use crate::erc721::core::Core;
+use crate::erc721::erc721::errors::Error::{NotAnOwnerOrApproved, ZeroAddress};
+use crate::erc721::erc721::Erc721;
+use crate::erc721::extensions::erc721metadata::{Erc721Metadata, Metadata};
+use crate::erc721token::errors::Error;
+
+#[odra::module]
+pub struct Erc721Token {
+    // ERC 721 core functions
+    core: Core,
+
+    // metadata field using erc721metadata trait
+    metadata: Metadata
+}
+
+#[odra::module]
+impl Erc721MM for Erc721Token {
+    #[odra(init)]
+    pub fn init(&mut self, name: String, symbol: String, base_uri: String) {
+        self.metadata.init(name, symbol, base_uri);
+    }
+
+    pub fn name(&self) -> String {
+        self.metadata.name()
+    }
+
+    pub fn symbol(&self) -> String {
+        self.metadata.symbol()
+    }
+
+    pub fn base_uri(&self) -> String {
+        self.metadata.base_uri()
+    }
+
+    pub fn balance_of(&self, owner: Address) -> U256 {
+        self.core.balance_of(owner)
+    }
+
+    pub fn owner_of(&self, token_id: U256) -> Address {
+        self.core.owner_of(token_id)
+    }
+
+    pub fn safe_transfer_from(&mut self, from: Address, to: Address, token_id: U256) {
+        self.core.safe_transfer_from(from, to, token_id);
+    }
+
+    pub fn transfer_from(&mut self, from: Address, to: Address, token_id: U256) {
+        self.core.transfer_from(from, to, token_id);
+    }
+
+    pub fn approve(&mut self, approved: Address, token_id: U256) {
+        self.core.approve(approved, token_id);
+    }
+
+    pub fn set_approval_for_all(&mut self, operator: Address, approved: bool) {
+        self.core.set_approval_for_all(operator, approved);
+    }
+
+    pub fn get_approved(&self, token_id: U256) -> Option<Address> {
+        self.core.get_approved(token_id)
+    }
+
+    pub fn is_approved_for_all(&self, owner: Address, operator: Address) -> bool {
+        self.core.is_approved_for_all(owner, operator)
+    }
+
+    pub fn mint(&mut self, to: Address, token_id: U256) {
+        if self.core.exists(&token_id) {
```

Comment:
Everyone can mint and burn. Maybe use `Ownable` to guard mint and burn? And the name to `OwnedERC721Metadata` or sth?

Reply:
How about now?